File: mini-qpu.py
import math
import random

class QPU:
    def __init__(self, num_qubits):
        self.num_qubits = num_qubits
        self.num_terms = 1 << num_qubits
        self._sv = [0.0] * self.num_terms
        self._sv[0] = 1.0

    def cnot(self, target, conditions):
        for t1 in range(self.num_terms):
            if t1 & target and (t1 & conditions) == conditions:
                t0 = t1 ^ target
                a,b = self._sv[t0],self._sv[t1]
                self._sv[t0] = b
                self._sv[t1] = a

    def had(self, target):
        r2 = 1.0 / math.sqrt(2.0)
        for t1 in range(self.num_terms):
            if t1 & target:
                t0 = t1 ^ target
                a,b = self._sv[t0],self._sv[t1]
                self._sv[t0] = (a + b) * r2
                self._sv[t1] = (a - b) * r2

    def phase(self, theta_degrees, conditions):
        rads = math.pi * theta_degrees / 180.0
        rot = math.cos(rads) + 1j*math.sin(rads)
        for t1 in range(self.num_terms):
            if (t1 & conditions) == conditions:
                self._sv[t1] *= rot

    def read(self, target):
        prob = 0.0
        for t1 in range(self.num_terms):
            if t1 & target:
                b = self._sv[t1]
                prob += b.real * b.real + b.imag * b.imag
        if random.random() <= prob:
            result = target
            scale = 1.0 / math.sqrt(1.0 - prob)
        else:
            result = 0
            scale = 1.0 / math.sqrt(prob)
        for t1 in range(self.num_terms):
            if ((t1 ^ result) & target):
                self._sv[t1] = 0.0
            else:
                self._sv[t1] *= scale
        return result


def run_test():
    qc = QPU(2)
    print(qc._sv)
    qc.had(1)
    qc.cnot(2, 1)
    qc.phase(90, 1)
    print(qc._sv)
    print('read',qc.read(1))
    print(qc._sv)
    # for i in range(10):
    #     qc.had(1 << i)
    #     qc.phase(10, 1 << i)
    # print(qc._sv)


run_test()

File: mini-qpu.cpp
// To build this: gcc mini-qpu.cpp -lstdc++
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include <vector>
#include <complex>
typedef std::complex<float> complex;

class QPU {
public:
    QPU(int _num_qubits) {
        num_qubits = _num_qubits;
        num_terms = 1 << num_qubits;
        _sv.resize(num_terms);
        _sv[0] = 1.0;
    }
    ~QPU() {}

    void cnot(int target, int conditions) {
        for (int t1 = 0; t1 < num_terms; ++t1) {
            if (t1 & target && (t1 & conditions) == conditions) {
                int t0 = t1 ^ target;
                complex a = _sv[t0];
                complex b = _sv[t1];
                _sv[t0] = b;
                _sv[t1] = a;
            }
        }
    }

    void had(int target) {
        float r2 = 1.0 / sqrt(2.0);
        for (int t1 = 0; t1 < num_terms; ++t1) {
            if (t1 & target) {
                int t0 = t1 ^ target;
                complex a = _sv[t0];
                complex b = _sv[t1];
                _sv[t0] = (a + b) * r2;
                _sv[t1] = (a - b) * r2;
            }
        }
    }

    void phase(float theta_degrees, int conditions) {
        float rads = M_PI * theta_degrees / 180.0;
        complex rot = complex(cos(rads), sin(rads));
        for (int t1 = 0; t1 < num_terms; ++t1) {
            if ((t1 & conditions) == conditions)
                _sv[t1] *= rot;
        }
    }

    int read(int target) {
        float prob = 0.0;
        for (int t1 = 0; t1 < num_terms; ++t1) {
            if (t1 & target) {
                complex b = _sv[t1];
                prob += b.real() * b.real() + b.imag() * b.imag();
            }
        }
        float scale;
        int result;
        if (((float) rand() / (RAND_MAX)) <= prob) {
            result = target;
            scale = 1.0 / sqrt(1.0 - prob);
        } else {
            result = 0;
            scale = 1.0 / sqrt(prob);
        }
        for (int t1 = 0; t1 < num_terms; ++t1) {
            if ((t1 ^ result) & target)
                _sv[t1] = 0.0;
            else
                _sv[t1] *= scale;
        }
        return result;
    }

    void print_sv() {
        for (int t1 = 0; t1 < num_terms; ++t1)
            printf("(%f,%f), ", _sv[t1].real(), _sv[t1].imag());
        printf("\n");
    }

    int num_qubits;
    int num_terms;
protected:
    std::vector<complex> _sv;
};



int main() {
    srand(time(NULL));

    QPU qc(2);
    qc.print_sv();
    qc.had(1);
    qc.cnot(2, 1);
    qc.phase(90, 1);
    qc.print_sv();
    printf("read %d\n", qc.read(1));
    qc.print_sv();
}
